File: scala/src/P01PureFunction.scala
object P01PureFunction {

  def impure1(x: Int): Int = {
    print(s"x=${x}")
    x
  }

  def impure2(x: Int): Double = {
    x * Math.random()
  }

  def impure3(x: Int): Int = {
    //        x = 4
    x
  }

  def impure4(x: Int) = {
    print(x + 3)
  }

  def impure5(): Int = {
    x + 1
  }

  def impure6(x: Int): Int = {
    var y = x + 1
    y
  }

  val x = 4

  val add1: Int => Int = y => y + x

  def add1f(y: Int): Int = y + x

  val add2: Int => Int = ((y: Int) => (x: Int) => y + x) (3)

  def add2f(y: Int): Int = ((x: Int) => y + x) (3)

  def add2ff(y: Int): Int = y + 3

  val add3: Int => Int = (w: Int) => ((y: Int) => ((x: Int) => y + x) (1)) (2)

  def add3f(w: Int): Int = ((y: Int) => ((x: Int) => y + x) (1)) (2)

  def add3ff(w: Int) = ((x: Int) => 2 + x) (1)

  def add3fff(w: Int) = 3

  def add3ffff() = 3

  def main(args: Array[String]): Unit = {
    assert(add1(1) == 5)
    assert(add1f(1) == 5)

    assert(add2(1) == 4)
    assert(add2f(1) == 4)
    assert(add2ff(1) == 4)

    assert(add3(1) == 3)
    assert(add3f(1) == 3)
    assert(add3ff(1) == 3)
    assert(add3fff(1) == 3)
    assert(add3ffff() == 3)

    assert(add3(2) == 3)
    assert(add3f(2) == 3)
  }
}

File: java/src/P01PureFunction.java
import java.util.function.Function;

public class P01PureFunction {

    private Integer impure1(Integer x) {
        System.out.println("x=" + x);
        return x;
    }

    private Double impure2(Integer x) {
        return x * Math.random();
    }

    private Integer impure3(Integer x) {
        x = 4;
        return x;
    }

    private void impure4(Integer x) {
        System.out.println(x + 3);
    }

    private Integer y = 1;

    private Integer impure5() {
        return y + 1;
    }

    private Integer impure6(Integer x) {
        var y = x + 1;
        return y;
    }

    private static Integer x = 4;

    private static Function<Integer, Integer> add1 = y -> y + x;

    private static Integer add1f(Integer y) {
        return y + x;
    }

    private static Function<Integer, Integer> add2 = ((Function<Integer, Function<Integer, Integer>>) y -> x -> y + x).apply(3);

    private static Integer add2f(Integer y) {
        return ((Function<Integer, Integer>) x -> y + x).apply(3);
    }

    private static Integer add2ff(Integer y) {
        return y + 3;
    }

    private static Function<Integer, Integer> add3 =
            w -> ((Function<Integer, Integer>) y -> ((Function<Integer, Integer>) x -> y + x).apply(1)).apply(2);

    private static Integer add3f(Integer w) {
        return ((Function<Integer, Integer>) y -> (((Function<Integer, Integer>) x -> y + x).apply(1))).apply(2);
    }

    private static Integer add3ff(Integer w) {
        return ((Function<Integer, Integer>) x -> 2 + x).apply(1);
    }

    private static Integer add3fff(Integer w) {
        return 3;
    }

    private static Integer add3ffff() {
        return 3;
    }

    public static void main(String[] args) {
        assert 5 == add1.apply(1);
        assert 5 == add1f(1);

        assert 4 == add2.apply(1);
        assert 4 == add2f(1);
        assert 4 == add2ff(1);

        assert 3 == add3.apply(1);
        assert 3 == add3f(1);
        assert 3 == add3ff(1);
        assert 3 == add3fff(1);
        assert 3 == add3ffff();

        assert 3 == add3.apply(2);
        assert 3 == add3f(2);
    }
}